src/tools/wasm-autogen.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2021 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Auto generate various useful code.
+//
+
+#include "support/command-line.h"
+#include "wasm.h"
+
+using namespace wasm;
+
+namespace {
+
+// Emit a declaration. If "startOfImpl" is true then we emit the beginning of
+// the implementation, which is identical to the declaration except it has no
+// ";" at the end.
+template<typename T> void autogenOneCAPIDecl(bool startOfImpl = false) {
+  MixedArena allocator;
+  T curr(allocator);
+
+  std::cout << '\n';
+
+#define DELEGATE_ID curr._id
+
+#define DELEGATE_START(id)                                                     \
+  std::cout << "BINARYEN_API BinaryenExpressionRef\n"                          \
+            << "Binaryen" << #id << "(BinaryenModuleRef module";
+
+#define DELEGATE_FIELD_CHILD(id, name)                                         \
```

Comment:
why are these macros? looks like some inline functions would not be that much more verbose?

Reply:
This lets them all refer to their shared context, which is useful because theses are usually used to generate fragments of a single larger function.